### PR TITLE
ISSUE-303 RateLimitingPlanUserRepository API and InMemory + contract

### PR DESCRIPTION
**Pull request by quantranhong1999**
https://github.com/linagora/james-project-private/issues/303

### CODE REVIEW

File: tmail-backend/pom.xml [hunk 1]
```diff
@@ -48,6 +48,7 @@
         <module>jmap/extensions-cassandra</module>
         <module>jmap/extensions-es-v7</module>
 
+        <module>rate-limit</module>
```
Comment (chibenwa) [inline tmail-backend/pom.xml:51]: <module>rate-limit</module>

=> 

<module>rate-limit/rate-limit-api</module>

(We will need multi-modules EG with Cassandra implementation)

File: tmail-backend/rate-limit/pom.xml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.projectreactor</groupId>
+            <artifactId>reactor-core</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.projectreactor</groupId>
+            <artifactId>reactor-scala-extensions_2.13</artifactId>
+            <version>0.8.0</version>
```
Comment (chibenwa) [inline tmail-backend/rate-limit/pom.xml:34]: Why do we need to specify this version?

File: tmail-backend/rate-limit/src/main/scala/com/linagora/tmail/rate/limit/model/RateLimitation.scala [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package com.linagora.tmail.rate.limit.model
+
+import org.apache.james.core.Username
+
+case class RateLimitingPlanId(value: Int)
```
Comment (chibenwa) [inline tmail-backend/rate-limit/src/main/scala/com/linagora/tmail/rate/limit/model/RateLimitation.scala:5]: Please use a UUID as an Id

File: tmail-backend/rate-limit/src/main/scala/com/linagora/tmail/rate/limit/plan/api/RateLimitingPlanUserRepository.scala [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package com.linagora.tmail.rate.limit.plan.api
+
+import com.linagora.tmail.rate.limit.model.RateLimitingPlanId
+import org.apache.james.core.Username
+import org.reactivestreams.Publisher
+
+trait RateLimitingPlanUserRepository {
+  def create(username: Username, planId: RateLimitingPlanId): Publisher[Unit]
```
Comment (chibenwa) [inline tmail-backend/rate-limit/src/main/scala/com/linagora/tmail/rate/limit/plan/api/RateLimitingPlanUserRepository.scala:8]: create => applyPlan ?

File: tmail-backend/rate-limit/src/main/scala/com/linagora/tmail/rate/limit/plan/api/RateLimitingPlanUserRepository.scala [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+  def delete(username: Username): Publisher[Unit]
```
Comment (chibenwa) [inline tmail-backend/rate-limit/src/main/scala/com/linagora/tmail/rate/limit/plan/api/RateLimitingPlanUserRepository.scala:10]: delete => revokePlan ?

File: tmail-backend/rate-limit/src/main/scala/com/linagora/tmail/rate/limit/plan/api/RateLimitingPlanUserRepository.scala [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+  def listUser(planId: RateLimitingPlanId): Publisher[Username]
```
Comment (chibenwa) [inline tmail-backend/rate-limit/src/main/scala/com/linagora/tmail/rate/limit/plan/api/RateLimitingPlanUserRepository.scala:12]: What is this needed for? Can we remove this method for now?

Comment (quantranhong1999) [reply tmail-backend/rate-limit/src/main/scala/com/linagora/tmail/rate/limit/plan/api/RateLimitingPlanUserRepository.scala:12]: image
This should help the admin easier to manage Plan-users mapping?

Comment (chibenwa) [reply tmail-backend/rate-limit/src/main/scala/com/linagora/tmail/rate/limit/plan/api/RateLimitingPlanUserRepository.scala:12]: ok

File: tmail-backend/rate-limit/rate-limit-api/pom.xml [hunk 7]
```diff
@@ -0,0 +1,41 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (vttranlina) [inline tmail-backend/rate-limit/rate-limit-api/pom.xml:1]: rename `rate-limit` ->  `rate-limiter` 
will similar with module name in James

File: tmail-backend/rate-limiter/rate-limiter-api/src/main/scala/com/linagora/tmail/rate/limiter/api/RateLimitingPlanUserRepository.scala [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+import com.linagora.tmail.rate.limiter.model.RateLimitingPlanId
+import org.apache.james.core.Username
+import org.reactivestreams.Publisher
+
+trait RateLimitingPlanUserRepository {
+  def applyPlan(username: Username, planId: RateLimitingPlanId): Publisher[Unit]
+
+  def revokePlan(username: Username): Publisher[Unit]
+
+  def listUser(planId: RateLimitingPlanId): Publisher[Username]
```
Comment (Arsnael) [inline tmail-backend/rate-limiter/rate-limiter-api/src/main/scala/com/linagora/tmail/rate/limiter/api/RateLimitingPlanUserRepository.scala:12]:
```suggestion
  def listUsers(planId: RateLimitingPlanId): Publisher[Username]
```

**Review by Arsnael — COMMENTED** [review]
Just a little remark, but overall good :)

**Review by Arsnael — APPROVED** [review]

**Review by chibenwa — APPROVED** [review]
Can you rebase?

**Pull request merged**